odl/solvers/scalar/steplen.py

```diff
@@ -148,6 +148,21 @@ def __call__(self, x, direction, dir_derivative):
             The computed step length
         """
         alpha = 1.0
+
+        # Since all comparisons with nan returns False, in order to not get a
+        # problem with nan-values we throw an error.
+        if self.function(x) == float('nan'):
+            raise NameError('ObjFunIsNanError')
```

Comment:
`NameError` [is the wrong exception](https://docs.python.org/3.5/library/exceptions.html#NameError) for this case. It's raised when you refer to a name in the code which has never been bound to a value. Like
```python
>>> u  # in a fresh console
NameError: name 'u' is not defined
```
The correct exception is `ValueError` with a reasonable message like `'function value is NaN, comparison for line search impossible.'`.

The more interesting issue is how to handle this issue one level above. Should the method bail out or rather try some new point `x` to get out of the NaN regime?

In any case, the docstring of this line search class should clearly state that this exception is raised if the function value NaN is occurred. Add a section like this:
```
Raises
------
ValueError
    if the function value at ``x`` is NaN
```